SmartTutor/ContentModel/LearningObjects/ChallengeModel/FulfillmentStrategy/BasicMetricsChecker.cs

```diff
@@ -29,20 +33,49 @@ private List<CaDETMember> GetMethodsFromClasses(List<CaDETClass> caDETClasses)
 
         private bool ValidateClassMetricRules(List<CaDETClass> caDETClasses)
         {
+            int flag = 0;
+            _metricHints = new List<MetricHint>();
             foreach (CaDETClass caDETClass in caDETClasses)
+            {
                 foreach (MetricRangeRule classMetricRule in _classMetricRules)
+                {
                     if (!classMetricRule.MetricMeetsRequirements(caDETClass.Metrics))
-                        return false;
-            return true;
+                    {
+                        flag++;
+                        AddHintForUnfulfilledMetricRule(classMetricRule);
+                    }
+                }
+            }
+            if (flag == 0)
+                return true;
+            return false;
         }
 
         private bool ValidateMethodMetricRules(List<CaDETMember> caDETMethods)
         {
+            int flag = 0;
             foreach (CaDETMember caDETMethod in caDETMethods)
+            {
                 foreach (MetricRangeRule methodMetricRule in _methodMetricRules)
+                {
                     if (!methodMetricRule.MetricMeetsRequirements(caDETMethod.Metrics))
-                        return false;
-            return true;
+                    {
+                        flag++;
+                        AddHintForUnfulfilledMetricRule(methodMetricRule);
+                    }
+                }
+            }
+            if (flag == 0)
+                return true;
+            return false;
+        }
+
+        private void AddHintForUnfulfilledMetricRule(MetricRangeRule metricRule)
+        {
+            _metricHints.Add(new MetricHint
+            {
+                Content = "Metric rule " + metricRule.MetricName + " should be between " + metricRule.FromValue + " and " + metricRule.ToValue + "."
```

Comment:
This isn't really a hint so much as the solution :)

I've been thinking about a more generalized ChallengeHint class that might even include a learning object. Basically, it would have the Content field like now + a reference to a LearningObject it might load (as a reminder). Maybe we should have the abstract strategy class contain a list of ChallengeHints and the implementation can decide the rules for when to return a specific hint. This will enable the trainee to ask for hints before they submit a challenge.

Reply:
Understood! All minor questions I will ask tomorrow.